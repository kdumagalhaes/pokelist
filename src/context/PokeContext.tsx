import {
  createContext,
  ReactNode,
  useContext,
  useState,
  useEffect,
  useCallback,
} from 'react'

interface Pokemon {
  id: number
  name: string
  url: string
}

interface PokeProviderProps {
  children: ReactNode
}

interface FavoritePokemon {
  id: number
  name: string
}
interface PokeContextModel {
  apiResponse: Pokemon[]
  getPokePicture: (id: string) => string
  nextPageUrl: string
  prevPageUrl: string
  goToNextPage: () => void
  goToPrevPage: () => void
  getPokeFavorites: (favoritePokemon: FavoritePokemon) => void
  favoritesList: FavoritePokemon[]
}

export const PokeContext = createContext({} as PokeContextModel)

export const PokeProvider = ({ children }: PokeProviderProps) => {
  const [apiResponse, setApiResponse] = useState<Pokemon[]>([])
  const [favoritesList, setFavoritesList] = useState<FavoritePokemon[]>([])
  const [currentPage, setCurrentPage] = useState(
    'https://pokeapi.co/api/v2/pokemon',
  )
  const [nextPageUrl, setNextPageUrl] = useState('')
  const [prevPageUrl, setPrevPageUrl] = useState('')

  // retrieve the basic pokemon data in the general list and pagination url
  useEffect(() => {
    fetch(currentPage)
      .then((res) => res.json())
      .then((json) => {
        setApiResponse(json.results)
        setNextPageUrl(json.next)
        setPrevPageUrl(json.previous)
      })
  }, [currentPage])

  useEffect(() => {
    const favoritesValue = localStorage.getItem('pokefavs')
    if (typeof favoritesValue === 'string') {
      const favorites = JSON.parse(favoritesValue)
      setFavoritesList(favorites)
    }
  }, [])

  const goToNextPage = () => {
    setCurrentPage(nextPageUrl)
  }

  const goToPrevPage = () => {
    setCurrentPage(prevPageUrl)
  }

  // retrieve id and name of the pokemon that the user has clicked on the fav button and save the data in localStore
  const getPokeFavorites = (favoritePokemon: FavoritePokemon) => {
    // if favoritesList doesn't have a pokemon with the same id as favoritePokemon, include it in the list
    if (!favoritesList.find((pokemon) => pokemon.id === favoritePokemon.id)) {
      localStorage.setItem(
        'pokefavs',
        JSON.stringify([...favoritesList, favoritePokemon]),
      )
    }
  }

  // retrieve the pokemon picture based on its id, which is presented in the general pokemon list
  const getPokePicture = useCallback((id: string) => {
    return `https://raw.githubusercontent.com/PokeAPI/sprites/master/sprites/pokemon/other/official-artwork/${id}.png`
  }, [])

  const value = {
    apiResponse,
    getPokePicture,
    nextPageUrl,
    prevPageUrl,
    goToNextPage,
    goToPrevPage,
    getPokeFavorites,
    favoritesList,
  }

  return <PokeContext.Provider value={value}>{children}</PokeContext.Provider>
}

export const usePokemon = () => {
  const context = useContext(PokeContext)

  if (context === undefined) {
    throw new Error('usePokemon must be used within PokeContext')
  }

  return context
}
