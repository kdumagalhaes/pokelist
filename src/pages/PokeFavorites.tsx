export function PokeFavorites() {
  return <h1>favorites</h1>
}
