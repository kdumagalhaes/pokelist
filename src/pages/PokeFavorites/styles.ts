import styled from 'styled-components'

export const Container = styled.div`
  max-width: 54rem;
  margin: 3rem auto 15rem auto;
  display: flex;
  flex-direction: column;
  align-items: center;
`

export const Grid = styled.div`
  display: grid;
  grid-template-columns: 1fr 1fr 1fr;
  gap: 2rem;
  cursor: pointer;
  margin-bottom: 6rem;

  @media only screen and (max-width: 600px) {
    display: flex;
    flex-direction: column;
    align-items: center;
    padding: 0 2rem;
  }
`
