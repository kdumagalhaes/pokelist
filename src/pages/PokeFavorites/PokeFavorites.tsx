/* eslint-disable react/no-unescaped-entities */
import { usePokemon } from '../../context/PokeContext'
import { PokeCard } from '../../components/PokeCard/PokeCard'
import { Container, EmptyListMessage, Grid } from './styles'
import SadPikachu from '../../assets/images/sad-pikachu.webp'

export function PokeFavorites() {
  const { favoritesList, getPokePicture } = usePokemon()

  return (
    <Container>
      {!favoritesList.length ? (
        <EmptyListMessage>
          <img src={SadPikachu} alt="sad pikachu" className="sad-pikachu" />
          <p className="message">You don't have any favorite Pok√©mons yet!</p>
          <a href="/" className="catch-pokemon-link">
            Catch 'em all!
          </a>
        </EmptyListMessage>
      ) : (
        <Grid>
          {favoritesList.map((pokemon) => {
            return (
              <PokeCard
                key={pokemon.name}
                name={pokemon.name}
                image={getPokePicture(pokemon.id.toString())}
              />
            )
          })}
        </Grid>
      )}
    </Container>
  )
}
