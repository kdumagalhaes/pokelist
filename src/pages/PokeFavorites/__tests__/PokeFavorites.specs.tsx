import { render, screen } from '@testing-library/react'
import { BrowserRouter } from 'react-router-dom'
import { PokeFavorites } from '../PokeFavorites'
import { PokeProvider } from '../../../context/PokeContext'

describe('PokeFavorites', () => {
  it('initial component should match snapshot', () => {
    const { asFragment } = render(
      <BrowserRouter>
        <PokeProvider>
          <PokeFavorites />
        </PokeProvider>
      </BrowserRouter>,
    )
    expect(asFragment()).toMatchSnapshot()
  })

  it('should render pikachu picture', () => {
    render(
      <BrowserRouter>
        <PokeProvider>
          <PokeFavorites />
        </PokeProvider>
      </BrowserRouter>,
    )
    const image = screen.getByAltText('sad pikachu')
    expect(image).toBeInTheDocument()
  })

  it('should render the message', () => {
    render(
      <BrowserRouter>
        <PokeProvider>
          <PokeFavorites />
        </PokeProvider>
      </BrowserRouter>,
    )
    const message = screen.getByText(`You don't have any favorite PokÃ©mon yet!`)
    expect(message).toBeInTheDocument()
  })

  it('should render the button', () => {
    render(
      <BrowserRouter>
        <PokeProvider>
          <PokeFavorites />
        </PokeProvider>
      </BrowserRouter>,
    )
    const message = screen.getByText(`Catch 'em all!`)
    expect(message).toBeInTheDocument()
  })
})
