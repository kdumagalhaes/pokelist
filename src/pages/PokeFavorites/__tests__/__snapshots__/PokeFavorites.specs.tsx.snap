// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PokeFavorites initial component should match snapshot 1`] = `
<DocumentFragment>
  <div
    class="sc-gswNZR dmvrTK"
  >
    <div
      class="sc-hLBbgP iOEvxQ"
    >
      <img
        alt="sad pikachu"
        class="sad-pikachu"
        src="sad-pikachu.webp"
      />
      <p
        class="message"
      >
        You don't have any favorite Pok√©mons yet!
      </p>
      <button
        class="catch-pokemon-btn"
      >
        Catch 'em all!
      </button>
    </div>
  </div>
</DocumentFragment>
`;
